```
--- FAIL: Test_Flannel_ParseIP (0.00s)
panic: json: cannot unmarshal string into Go struct field Service.CheckAPI of type bool [recovered]
	panic: json: cannot unmarshal string into Go struct field Service.CheckAPI of type bool

goroutine 22 [running]:
testing.tRunner.func1(0xc0000e6600)
	/usr/local/go/src/testing/testing.go:792 +0x6a7
panic(0xb6dfe0, 0xc0000b6be0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/giantswarm/k8s-kvm-health/vendor/github.com/giantswarm/microkit/flag.Init(0xb2d440, 0xc000093080)
	/go/src/github.com/giantswarm/k8s-kvm-health/vendor/github.com/giantswarm/microkit/flag/flag.go:33 +0x4cb
github.com/giantswarm/k8s-kvm-health/flag.New(0xaee1bd)
	/go/src/github.com/giantswarm/k8s-kvm-health/flag/flag.go:14 +0x94
github.com/giantswarm/k8s-kvm-health/service.Test_Flannel_ParseIP.func1(0xc000168000, 0x61, 0x70, 0xc00014a280, 0x139, 0x140, 0xdf3fd0)
	/go/src/github.com/giantswarm/k8s-kvm-health/service/flannel_test.go:21 +0xc8
github.com/giantswarm/k8s-kvm-health/service.Test_Flannel_ParseIP(0xc0000e6600)
	/go/src/github.com/giantswarm/k8s-kvm-health/service/flannel_test.go:113 +0x54d
testing.tRunner(0xc0000e6600, 0xc24aa8)
	/usr/local/go/src/testing/testing.go:827 +0x163
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x651
FAIL	github.com/giantswarm/k8s-kvm-health/service	0.034s
?   	github.com/giantswarm/k8s-kvm-health/service/healthz	[no test files]
?   	github.com/giantswarm/k8s-kvm-health/service/healthz/kvm	[no test files]
```
any idea how could i go around this?
ping @JosephSalisbury

--- a/flag/service/service.go
+++ b/flag/service/service.go
@@ -1,7 +1,7 @@
 package service
 
 type Service struct {
-	CheckAPI      bool
+	CheckAPI      string
 	FlannelFile   string
 	ListenAddress string
 	IPAddress     string

--- a/main.go
+++ b/main.go
@@ -32,11 +32,7 @@ func readEnv() error {
 	// load conf from ENV
 	f.Service.FlannelFile = os.Getenv("NETWORK_ENV_FILE_PATH")
 	f.Service.ListenAddress = os.Getenv("LISTEN_ADDRESS")
-	if os.Getenv("CHECK_K8S_API") == "true" {
-		f.Service.CheckAPI = true
-	} else {
-		f.Service.CheckAPI = false
-	}
+	f.Service.CheckAPI = os.Getenv("CHECK_K8S_API")
 	if f.Service.FlannelFile == "" {
 		return microerror.Maskf(invalidConfigError, "NETWORK_ENV_FILE_PATH must not be empty")
 	}

--- a/service/healthz/kvm/kvm.go
+++ b/service/healthz/kvm/kvm.go
@@ -8,6 +8,7 @@ import (
 	"github.com/giantswarm/micrologger"
 	"github.com/sparrc/go-ping"
 	"net/http"
+	"net/url"
 	"time"
 )
 
@@ -19,7 +20,11 @@ const (
 	Name = "kvmHealthz"
 
 	// config
-	pingCount = 1
+	pingCount      = 1
+	httpsScheme    = "https"
+	httpScheme     = "http"
+	k8sAPIPort     = 443
+	k8sKubeletPort = 10248
 )
 
 // Config represents the configuration used to create a healthz service.
@@ -63,7 +68,9 @@ func New(config Config) (*Service, error) {
 
 // GetHealthz implements the health check for network interface.
 func (s *Service) GetHealthz(ctx context.Context) (healthz.Response, error) {
-	pingFailed, pingMsg := s.pingHealthCheck()
+	var apiFailed, kubeletFailed, pingFailed bool
+	var apiMsg, kubeletMsg, pingMsg string
+	pingFailed, pingMsg = s.pingHealthCheck()
 
 	response := healthz.Response{
 		Description: Description,
@@ -72,9 +79,16 @@ func (s *Service) GetHealthz(ctx context.Context) (healthz.Response, error) {
 		Name:        Name,
 	}
 
-	// check api only if ping succeeded
-	if !pingFailed && s.checkAPI {
-		apiFailed, apiMsg := s.apiHealthCheck()
+	// check kubelet only if ping succeeded
+	if !pingFailed {
+		kubeletFailed, kubeletMsg = s.httpHealthCheck(k8sKubeletPort, httpScheme)
+		response.Failed = kubeletFailed
+		response.Message += kubeletMsg
+	}
+
+	// check api only if ping and kubelet succeeded
+	if !pingFailed && !kubeletFailed && s.checkAPI {
+		apiFailed, apiMsg = s.httpHealthCheck(k8sAPIPort, httpsScheme)
 		response.Failed = apiFailed
 		response.Message += apiMsg
 	}
@@ -110,18 +124,22 @@ func (s *Service) pingHealthCheck() (bool, string) {
 	return failed, message
 }
 
-// implementation of the interface healthz logic for k8s api check
-func (s *Service) apiHealthCheck() (bool, string) {
+// implementation of the interface healthz logic for http  check
+func (s *Service) httpHealthCheck(port int, scheme string) (bool, string) {
 	var message string
-	url := fmt.Sprintf("https://%s/healthz", s.ip)
-	// send request to k8s API
-	_, err := http.Get(url)
+	u := url.URL{
+		Host:   fmt.Sprintf("%s:%d", s.ip, port),
+		Path:   "healthz",
+		Scheme: scheme,
+	}
+	// send request to http endpoint
+	_, err := http.Get(u.String())
 	if err != nil {
-		message = fmt.Sprintf("Failed to send http request to k8s API. %s", err)
+		message = fmt.Sprintf("Failed to send http request to endpoint %s. %s", u.String(), err)
 		return true, message
 	}
 
-	message = fmt.Sprintf("Healthcheck for k8s API has been successful. K8s is live and responding. on %s.", url)
+	message = fmt.Sprintf("Healthcheck for http endpoint %s has been successful.", u.String())
 	// exit
 	return false, message
 }

--- a/service/healthz/kvm/kvm.go
+++ b/service/healthz/kvm/kvm.go
@@ -113,7 +113,7 @@ func (s *Service) pingHealthCheck() (bool, string) {
 // implementation of the interface healthz logic for k8s api check
 func (s *Service) apiHealthCheck() (bool, string) {
 	var message string
-	url := fmt.Sprintf("https://%s", s.ip)
+	url := fmt.Sprintf("https://%s/healthz", s.ip)
 	// send request to k8s API
 	_, err := http.Get(url)
 	if err != nil {

--- a/service/healthz/kvm/kvm.go
+++ b/service/healthz/kvm/kvm.go
@@ -66,7 +66,11 @@ func New(config Config) (*Service, error) {
 	return newService, nil
 }
 
-// GetHealthz implements the health check for network interface.
+// GetHealthz Provides Healthz implementation to check health status of network
+// interface. It performs following checks in given order:
+//  - Ping configured IP. 
+//  - Check that Kubelet instance in configured IP responds to HTTP request.
+//  - Check that K8s API in configured IP responds to HTTPS request. 
 func (s *Service) GetHealthz(ctx context.Context) (healthz.Response, error) {
 	var apiFailed, kubeletFailed, pingFailed bool
 	var apiMsg, kubeletMsg, pingMsg string

--- a/service/service.go
+++ b/service/service.go
@@ -62,7 +62,11 @@ func New(config Config) (*Service, error) {
 	{
 		healthzConfig := healthz.DefaultConfig()
 
-		healthzConfig.CheckAPI = config.Flag.Service.CheckAPI
+		if config.Flag.Service.CheckAPI == "true" {
+			healthzConfig.CheckAPI = true
+		} else {
+			healthzConfig.CheckAPI = false
+		}
 		healthzConfig.IPAddress = config.Flag.Service.IPAddress
 		healthzConfig.Logger = config.Logger
 

